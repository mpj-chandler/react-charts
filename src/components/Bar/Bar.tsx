import React from 'react';
import { AxisRange, Padding, SeriesDataPoint } from '../BarChart/types';
import { useAnimation } from '../../utils/useAnimation';
import Placement from '../../enums/Placement';
import { getXAxisRange } from '../../utils/getXAxisRange';

export interface BarProps {
    point: SeriesDataPoint
    yRange: AxisRange;
    xRange: AxisRange;
    width: number;
    fill: string;
    stroke: string;
    padding: Padding;
    numBars: number;
    barAlignment: Placement;
}

function getBarEndingHeight(props: BarProps): number {
    const yMin = Math.min(0, Math.max(0, props.yRange.min));
    
    if (props.yRange.max <= 0) {
        return (100 - props.padding.top) * ((yMin - props.point.y) / (props.yRange.min - props.yRange.max));
    }

    return (100 - props.padding.top) * ((props.point.y - yMin) / (props.yRange.max - props.yRange.min));
}

function getZeroPoint(props: BarProps): number {
    return (props.yRange.max / (props.yRange.max - props.yRange.min));
}

function getBarXPosition(props: BarProps) {
    const interval = props.xRange.max - props.xRange.min;

    if (props.barAlignment === Placement.Bucket) {
        const step = 0.5 * interval / props.numBars;
        return (100 - props.padding.right) * ((step + props.point.x) / interval);
    }
    return (100 - props.padding.right) * (props.point.x / interval);
}

function getY(height: number, yStart: number, value: number) {
    
    if (value >= 0) {
        return yStart - height;
    }

    return yStart;
}

const Bar: React.FC<BarProps> = (props: BarProps) => {
    const x = getBarXPosition(props);
    const height = getBarEndingHeight(props);
    const zeroPoint = getZeroPoint(props);
    const animation = useAnimation('elastic', 600, 0);
    const yStart = props.yRange.max <= 0 ? props.padding.top : 100 * zeroPoint;
    
    return (
        <rect 
            x={`${x - props.width / 2}%`}
            y={`${getY(animation * height, yStart, props.point.y)}%`}
            height={`${props.point.y < 0 ? -1 * animation * height : animation * height}%`}
            width={`${props.width}%`}
            fill={props.fill}
            stroke={props.stroke}
        >
        </rect>
    )
}

export default Bar;
