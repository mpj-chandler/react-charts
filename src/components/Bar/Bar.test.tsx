import React from 'react';
import TestRenderer from 'react-test-renderer';
import Bar, { BarProps } from './Bar';
import Placement from '../../enums/Placement';
import DataType from '../../enums/DataType';

jest.useFakeTimers();

describe('Bar', () => {
    describe('When dealing with numeric data types', () => {
        describe('When all values on the yAxis are positive', () => {
            const props: BarProps = {
                point: { x: 10, y: 100 },
                index: 1,
                type: DataType.NonNullNumeric,
                xRange: { min: 0, max: 100 },
                yRange: { min: 0, max: 200 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                barAlignment: Placement.Aligned,
                numBars: 3,
            };

            it('it renders consistently with defaults', () => {
                const component = TestRenderer.create(<Bar {...props}/>);

                jest.runAllTimers();

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Aligned', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Aligned}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Bucket', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Bucket}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });
        });

        describe('When all values on the yAxis are negative', () => {
            const props: BarProps = {
                point: { x: 10, y: -50 },
                index: 1,
                type: DataType.NonNullNumeric,
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: -100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                barAlignment: Placement.Aligned,
                numBars: 3,
            };

            it('it renders consistently with defaults', () => {
                const component = TestRenderer.create(<Bar {...props}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Aligned', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Aligned}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Bucket', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Bucket}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });
        });


        describe('When the yAxis spans positive and negative values', () => {
            const props: BarProps = {
                index: 1,
                type: DataType.NonNullNumeric,
                point: { x: 10, y: -50 },
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: 100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                barAlignment: Placement.Aligned,
                numBars: 3,
            };

            it('it renders consistently with defaults', () => {
                const component = TestRenderer.create(<Bar {...props}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Aligned', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Aligned}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Bucket', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Bucket}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });
        });
    });

    describe('When dealing with named data types', () => {
        describe('When all values on the yAxis are positive', () => {
            const props: BarProps = {
                point: { x: 'Anything', y: 100 },
                index: 1,
                type: DataType.Named,
                xRange: { min: 0, max: 100 },
                yRange: { min: 0, max: 200 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                barAlignment: Placement.Aligned,
                numBars: 3,
            };

            it('it renders consistently with defaults', () => {
                const component = TestRenderer.create(<Bar {...props}/>);

                jest.runAllTimers();

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Aligned', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Aligned}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Bucket', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Bucket}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });
        });

        describe('When all values on the yAxis are negative', () => {
            const props: BarProps = {
                point: { x: 'Anything', y: -50 },
                index: 1,
                type: DataType.Named,
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: -100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                barAlignment: Placement.Aligned,
                numBars: 3,
            };

            it('it renders consistently with defaults', () => {
                const component = TestRenderer.create(<Bar {...props}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Aligned', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Aligned}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Bucket', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Bucket}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });
        });


        describe('When the yAxis spans positive and negative values', () => {
            const props: BarProps = {
                index: 1,
                type: DataType.Named,
                point: { x: 'Anything', y: -50 },
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: 100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                barAlignment: Placement.Aligned,
                numBars: 3,
            };

            it('it renders consistently with defaults', () => {
                const component = TestRenderer.create(<Bar {...props}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Aligned', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Aligned}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });

            it('it renders consistently with Placement set to Bucket', () => {
                const component = TestRenderer.create(<Bar {...props} barAlignment={Placement.Bucket}/>);

                expect(component.toJSON()).toMatchSnapshot();
            });
        });
    });

    describe('When attempting a render with an unknown data type', () => {
        it('it throws an error', () => {
            const props: BarProps = {
                index: 1,
                type: DataType.DateIndexed,
                point: { x: 'Anything', y: -50 },
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: 100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                barAlignment: Placement.Aligned,
                numBars: 3,
            };

            expect(() => TestRenderer.create(<Bar {...props} barAlignment={Placement.Aligned}/>)).toThrowError('Error in rendering bar! Unrecognized data type!');
        });
    });
});

