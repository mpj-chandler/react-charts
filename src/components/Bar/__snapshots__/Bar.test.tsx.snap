// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar When all values on the yAxis are negative it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-41%"
  y="10%"
/>
`;

exports[`Bar When all values on the yAxis are negative it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-26%"
  y="10%"
/>
`;

exports[`Bar When all values on the yAxis are negative it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-41%"
  y="10%"
/>
`;

exports[`Bar When all values on the yAxis are positive it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-41%"
  y="100%"
/>
`;

exports[`Bar When all values on the yAxis are positive it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-26%"
  y="100%"
/>
`;

exports[`Bar When all values on the yAxis are positive it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-41%"
  y="100%"
/>
`;

exports[`Bar When the yAxis spans positive and negative values it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-41%"
  y="50%"
/>
`;

exports[`Bar When the yAxis spans positive and negative values it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-26%"
  y="50%"
/>
`;

exports[`Bar When the yAxis spans positive and negative values it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-41%"
  y="50%"
/>
`;
